packages/react-router/src/utils.ts

```diff
@@ -209,6 +209,32 @@ export function isPlainArray(value: unknown): value is Array<unknown> {
   return Array.isArray(value) && value.length === Object.keys(value).length
 }
 
+/**
+ * This function will remove all undefined values from an object.
+ * It will also remove any undefined values from nested objects.
+ * It will not remove undefined values from arrays.
+ */
+export function deepRemoveUndefinedFromObject<T>(obj: T): T {
```

Comment:
can we, instead of this function, enhance `deepEqual` to ignore properties that are explicitly set to `undefined`?

Reply:
I assume you object properties yes? Not to strip out `undefined` as an array value, only as an object value, when comparing `[undefined, 'foo', { a: undefined }]` with `['bar', undefined, { a: undefined }]`.

I was hesitant to make such changes to the `deepEqual` function itself since that'd have an impact on the `router.matchRoute` and `router.commitLocation` methods, especially without understanding the conditions in which `undefined` would get either pass or fail the function.

It's why I introduced that function since it explicitly does a single job and has a narrow impact area (only the link component).

Reply:
@schiller-manuel thoughts on having the `deepEqual` function internally call the `deepRemoveUndefinedFromObject` function on both its inputs `a` and `b`?

Reply:
Yup, that's what'll work best here. I'll beef up those tests for the `deepRemoveUndefinedFromObject` function and push this.